Unicode support: in script API String class should work with unicode chars

As #1297 was generally complete, there's still an issue of String class API using 8-bit `char` type to return and append text characters.

The proposal is to adjust related functions and properties to work with `int` based type instead. This may be simply `int` for now, or a new type declared for "text character" as opposed to `char`=byte.

In AGS script `char` and `int` may be assigned to each other without explicit conversion (unlike int-float), therefore the existing scripts will continue to work.

The related members are:

- String.Chars[ ] (indexed attribute)
- String.AppendChar
- String.ReplaceCharAt

On a separate note, there's a question of String.Format's `%c` specifier. In C/C++ this means strictly 1byte-char, so this will work only for character codes <256 now. Need to figure out whether to add another one for printing unicode chars (int), or allow to use same specifier for unicode too. Technically, the script vm allows the latter, as `char`, `short` and `int` all are represented as 32-bit ints when passed as function arguments in ags script vm.

> added compiler option SCOPT_UTF8

interesting, once this is in I can add it in the compiler frontend.

The windows test though will fail at build for the following error:

> script_api.cpp.obj : error LNK2001: unresolved external symbol `_usetc` [`Solutions\Tests.App\Engine.App.Test.vcxproj`]

--- a/Common/script/cc_options.h
+++ b/Common/script/cc_options.h
@@ -27,6 +27,7 @@
 #define SCOPT_NOIMPORTOVERRIDE 0x20 // do not allow an import to be re-declared
 #define SCOPT_LEFTTORIGHT 0x40   // left-to-right operator precedance
 #define SCOPT_OLDSTRINGS  0x80   // allow old-style strings
+#define SCOPT_UTF8        0x100  // UTF-8 text mode
 
 extern void ccSetOption(int, int);
 extern int ccGetOption(int);

--- a/Compiler/script/cs_parser.cpp
+++ b/Compiler/script/cs_parser.cpp
@@ -13,8 +13,8 @@
 #include "script/script_common.h"
 #include "script/cc_error.h"
 #include "cc_variablesymlist.h"
-
 #include "fmem.h"
+#include "util/utf8.h"
 
 extern int currentline;
 
@@ -218,9 +218,14 @@ int cc_tokenize(const char*inpl, ccInternalList*targ, ccCompiledScript*scrip) {
             thissymbol.push_back(fmem_getc(iii));
             symlen++;
         }
-        if ((thissymbol[0] == '\'') && (thissymbol[2] == '\'')) {
-            // convert the character to its ASCII equivalent
-            thissymbol = std::to_string(thissymbol[1]);
+        if ((thissymbol[0] == '\'') && (thissymbol.back() == '\'')) {
+            int chr = 0;
+            if (ccGetOption(SCOPT_UTF8)) {
+                Utf8::GetChar(&thissymbol[1], thissymbol.size() - 2, &chr);
+            } else {
+                chr = thissymbol[1];
+            }
+            thissymbol = std::to_string(chr);
         }
         else if (thissymbol[0] == '\'') {
             cc_error("incorrectly terminated character constant");

--- a/Editor/AGS.Editor/Resources/agsdefns.sh
+++ b/Editor/AGS.Editor/Resources/agsdefns.sh
@@ -446,7 +446,7 @@ internalstring autoptr builtin managed struct String {
   /// Returns a new string with the specified string appended to this string.
   import String  Append(const string appendText);
   /// Returns a new string that has the extra character appended.
-  import String  AppendChar(char extraChar);
+  import String  AppendChar(int extraChar);
   import int     Contains(const string needle);   // $AUTOCOMPLETEIGNORE$
   /// Creates a copy of the string.
   import String  Copy();
@@ -455,7 +455,7 @@ internalstring autoptr builtin managed struct String {
   /// Returns a lower-cased version of this string.
   import String  LowerCase();
   /// Returns a new string, with the specified character changed.
-  import String  ReplaceCharAt(int index, char newChar);
+  import String  ReplaceCharAt(int index, int newChar);
   /// Returns a portion of the string.
   import String  Substring(int index, int length);
   /// Truncates the string down to the specified length by removing characters from the end.
@@ -487,7 +487,7 @@ internalstring autoptr builtin managed struct String {
   /// Converts the string to an integer.
   readonly import attribute int AsInt;
   /// Accesses individual characters of the string.
-  readonly import attribute char Chars[];
+  readonly import attribute int Chars[];
   /// Returns the length of the string.
   readonly import attribute int Length;
 };

--- a/Editor/AGS.Native/ScriptCompiler.cpp
+++ b/Editor/AGS.Native/ScriptCompiler.cpp
@@ -75,6 +75,7 @@ namespace AGS
 			  ccSetOption(SCOPT_LINENUMBERS, 1);
 			  ccSetOption(SCOPT_LEFTTORIGHT, game->Settings->LeftToRightPrecedence);
 			  ccSetOption(SCOPT_OLDSTRINGS, !game->Settings->EnforceNewStrings);
+			  ccSetOption(SCOPT_UTF8, game->UnicodeMode);
 
         if (exceptionToThrow == nullptr)
         {

--- a/Engine/ac/string.cpp
+++ b/Engine/ac/string.cpp
@@ -25,6 +25,7 @@
 #include "debug/debug_log.h"
 #include "script/runtimescriptvalue.h"
 #include "util/string_compat.h"
+#include "util/utf8.h"
 
 using namespace AGS::Common;
 
@@ -52,13 +53,19 @@ const char* String_Append(const char *thisString, const char *extrabit) {
     return CreateNewScriptString(buffer, false);
 }
 
-const char* String_AppendChar(const char *thisString, char extraOne) {
-    char *buffer = (char*)malloc(strlen(thisString) + 2);
-    sprintf(buffer, "%s%c", thisString, extraOne);
+const char* String_AppendChar(const char *thisString, int extraOne) {
+    size_t chw = 1;
+    char chr[Utf8::UtfSz + 1]{};
+    if (get_uformat() == U_UTF8)
+        chw = Utf8::SetChar(extraOne, chr, sizeof(chr));
+    else
+        chr[0] = extraOne;
+    char *buffer = (char*)malloc(strlen(thisString) + chw + 1);
+    sprintf(buffer, "%s%s", thisString, chr);
     return CreateNewScriptString(buffer, false);
 }
 
-const char* String_ReplaceCharAt(const char *thisString, int index, char newChar) {
+const char* String_ReplaceCharAt(const char *thisString, int index, int newChar) {
     size_t len = ustrlen(thisString);
     if ((index < 0) || ((size_t)index >= len))
         quit("!String.ReplaceCharAt: index outside range of string");
@@ -67,12 +74,17 @@ const char* String_ReplaceCharAt(const char *thisString, int index, char newChar
     int uchar = ugetc(thisString + off);
     size_t remain_sz = strlen(thisString + off);
     size_t old_sz = ucwidth(uchar);
-    size_t new_sz = sizeof(char); // TODO: support unicode char in API
-    size_t total_sz = off + remain_sz + new_sz - old_sz + 1;
+    size_t new_chw = 1;
+    char new_chr[Utf8::UtfSz + 1]{};
+    if (get_uformat() == U_UTF8)
+        new_chw = Utf8::SetChar(newChar, new_chr, sizeof(new_chr));
+    else
+        new_chr[0] = newChar;
+    size_t total_sz = off + remain_sz + new_chw - old_sz + 1;
     char *buffer = (char*)malloc(total_sz);
     memcpy(buffer, thisString, off);
-    usetc(buffer + off, newChar);
-    memcpy(buffer + off + new_sz, thisString + off + old_sz, remain_sz - old_sz + 1);
+    memcpy(buffer + off, new_chr, new_chw);
+    memcpy(buffer + off + new_chw, thisString + off + old_sz, remain_sz - old_sz + 1);
     return CreateNewScriptString(buffer, false);
 }
 
@@ -205,9 +217,9 @@ const char* String_UpperCase(const char *thisString) {
 }
 
 int String_GetChars(const char *texx, int index) {
-    if ((index < 0) || (index >= (int)strlen(texx)))
+    if ((index < 0) || (index >= ustrlen(texx)))
         return 0;
-    return texx[index];
+    return ugetat(texx, index);
 }
 
 int StringToInt(const char*stino) {

--- a/Engine/ac/string.h
+++ b/Engine/ac/string.h
@@ -28,8 +28,8 @@
 int String_IsNullOrEmpty(const char *thisString);
 const char* String_Copy(const char *srcString);
 const char* String_Append(const char *thisString, const char *extrabit);
-const char* String_AppendChar(const char *thisString, char extraOne);
-const char* String_ReplaceCharAt(const char *thisString, int index, char newChar);
+const char* String_AppendChar(const char *thisString, int extraOne);
+const char* String_ReplaceCharAt(const char *thisString, int index, int newChar);
 const char* String_Truncate(const char *thisString, int length);
 const char* String_Substring(const char *thisString, int index, int length);
 int String_CompareTo(const char *thisString, const char *otherString, bool caseSensitive);

--- a/Engine/script/script_api.cpp
+++ b/Engine/script/script_api.cpp
@@ -14,11 +14,13 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <allegro.h>
 #include "ac/game_version.h"
 #include "script/cc_error.h"
 #include "script/runtimescriptvalue.h"
 #include "script/script_api.h"
 #include "util/math.h"
+#include "util/utf8.h"
 
 namespace Math = AGS::Common::Math;
 
@@ -29,6 +31,7 @@ enum FormatParseResult
     kFormatParseLiteralPercent,
     kFormatParseArgInteger,
     kFormatParseArgFloat,
+    kFormatParseArgCharacter,
     kFormatParseArgString,
     kFormatParseArgPointer,
 
@@ -135,8 +138,10 @@ const char *ScriptSprintf(char *buffer, size_t buf_length, const char *format,
                 case 'u':
                 case 'x':
                 case 'X':
-                case 'c':
                     fmt_done = kFormatParseArgInteger;
+                    break;
+                case 'c':
+                    fmt_done = kFormatParseArgCharacter;
                     break;
                 case 'e':
                 case 'E':
@@ -185,11 +190,24 @@ const char *ScriptSprintf(char *buffer, size_t buf_length, const char *format,
                 // NOTE: snprintf is called with avail_outbuf + 1 here, because we let it use our reserved
                 // character for null-terminator, in case we are at the end of the buffer
                 *fmt_bufptr = 0; // terminate the format buffer, we are going to use it
-                if (fmt_done == kFormatParseArgInteger)
-                    snprintf_res = snprintf(out_ptr, avail_outbuf + 1, fmtbuf, GetArgInt(sc_args, varg_ptr, arg_idx));
-                else if (fmt_done == kFormatParseArgFloat)
-                    snprintf_res = snprintf(out_ptr, avail_outbuf + 1, fmtbuf, GetArgFloat(sc_args, varg_ptr, arg_idx));
-                else
+                switch (fmt_done)
+                {
+                case kFormatParseArgInteger:
+                    snprintf_res = snprintf(out_ptr, avail_outbuf + 1, fmtbuf, GetArgInt(sc_args, varg_ptr, arg_idx)); break;
+                case kFormatParseArgFloat:
+                    snprintf_res = snprintf(out_ptr, avail_outbuf + 1, fmtbuf, GetArgFloat(sc_args, varg_ptr, arg_idx)); break;
+                case kFormatParseArgCharacter:
+                {
+                    int chr = GetArgInt(sc_args, varg_ptr, arg_idx);
+                    char cbuf[Utf8::UtfSz + 1]{};
+                    if (get_uformat() == U_UTF8)
+                        Utf8::SetChar(chr, cbuf, Utf8::UtfSz);
+                    else
+                        cbuf[0] = chr;
+                    snprintf_res = snprintf(out_ptr, avail_outbuf + 1, "%s", cbuf);
+                    break;
+                }
+                case kFormatParseArgString:
                 {
                     const char *p = GetArgPtr(sc_args, varg_ptr, arg_idx);
                     // Do extra checks for %s placeholder
@@ -212,6 +230,11 @@ const char *ScriptSprintf(char *buffer, size_t buf_length, const char *format,
                         return "";
                     }
                     snprintf_res = snprintf(out_ptr, avail_outbuf + 1, fmtbuf, p);
+                    break;
+                }
+                case kFormatParseArgPointer:
+                    snprintf_res = snprintf(out_ptr, avail_outbuf + 1, fmtbuf, GetArgPtr(sc_args, varg_ptr, arg_idx)); break;
+                default: /* should not happen */ break;
                 }
 
                 arg_idx++;

--- a/Solutions/Tests.App/Engine.App.Test.vcxproj
+++ b/Solutions/Tests.App/Engine.App.Test.vcxproj
@@ -23,6 +23,8 @@
     <ClCompile Include="..\..\Common\libsrc\googletest\src\gtest_main.cc" />
     <ClCompile Include="..\..\Engine\script\script_api.cpp" />
     <ClCompile Include="..\..\Engine\test\scsprintf_test.cpp" />
+    <ClCompile Include="..\..\libsrc\allegro\src\allegro.c" />
+    <ClCompile Include="..\..\libsrc\allegro\src\unicode.c" />
   </ItemGroup>
   <PropertyGroup Label="Globals">
     <ProjectGuid>{E5EBFBA9-1617-412B-843E-682609C65100}</ProjectGuid>

--- a/Solutions/Tests.App/Engine.App.Test.vcxproj.filters
+++ b/Solutions/Tests.App/Engine.App.Test.vcxproj.filters
@@ -13,6 +13,12 @@
     <ClCompile Include="..\..\Engine\script\script_api.cpp">
       <Filter>Engine</Filter>
     </ClCompile>
+    <ClCompile Include="..\..\libsrc\allegro\src\unicode.c">
+      <Filter>libsrc\allegro</Filter>
+    </ClCompile>
+    <ClCompile Include="..\..\libsrc\allegro\src\allegro.c">
+      <Filter>libsrc\allegro</Filter>
+    </ClCompile>
   </ItemGroup>
   <ItemGroup>
     <Filter Include="Common">
@@ -23,6 +29,12 @@
     </Filter>
     <Filter Include="Engine">
       <UniqueIdentifier>{9b2939ca-1439-4aa3-8a1b-e897e750739f}</UniqueIdentifier>
+    </Filter>
+    <Filter Include="libsrc">
+      <UniqueIdentifier>{eb289ce3-1947-47b4-9025-0a91f10e81bd}</UniqueIdentifier>
+    </Filter>
+    <Filter Include="libsrc\allegro">
+      <UniqueIdentifier>{8fd77dd5-e05e-44c4-a6f6-f6bf78b72dcd}</UniqueIdentifier>
     </Filter>
   </ItemGroup>
 </Project>

